File: frontend/src/app/services/devis/devis.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Devis {
  id?: number;
  dateDevis: string;
  amount: number;
  status: string;
  clientId?: number;
  artisanId?: number;
}

// Add this interface for the DTO that matches the backend
export interface DevisRequestDTO {
  dateDevis: string;
  amount: number;
  devisStatus: string;
  clientId?: number;
  artisanId?: number;
}

@Injectable({
  providedIn: 'root'
})
export class DevisService {

  constructor(private http: HttpClient) { }

  private apiUrl = 'http://localhost:8081/api/devis';

  createDevis(devis: Devis): Observable<Devis> {
    return this.http.post<Devis>(`${this.apiUrl}/create`, devis);
  }

  getAllDevis(): Observable<Devis[]> {
    return this.http.get<Devis[]>(`${this.apiUrl}/allDevis`);
  }

  getDevisById(id: number): Observable<Devis> {
    return this.http.get<Devis>(`${this.apiUrl}/${id}`);
  }

  updateDevis(id: number, devis: Devis): Observable<Devis> {
    return this.http.put<Devis>(`${this.apiUrl}/${id}`, devis);
  }

  deleteDevis(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  sendDevis(devis: Devis): Observable<Devis> {
    // Convert Devis to DevisRequestDTO
    const devisRequestDTO: DevisRequestDTO = {
      dateDevis: devis.dateDevis,
      amount: devis.amount,
      devisStatus: devis.status,
      clientId: devis.clientId,
      artisanId: devis.artisanId
    };
    
    return this.http.post<Devis>(`${this.apiUrl}/send`, devisRequestDTO);
  }

  listDevisByClient(clientId: number): Observable<Devis[]> {
    return this.http.get<Devis[]>(`${this.apiUrl}/Client/${clientId}`);
  }

  listDevisByArtisan(artisanId: number): Observable<Devis[]> {
    return this.http.get<Devis[]>(`${this.apiUrl}/Artisan/${artisanId}`);
  }

  updateStatus(id: number, status: string): Observable<Devis> {
    return this.http.put<Devis>(`${this.apiUrl}/${id}/status?status=${status}`, {});
  }
}

File: backend/herfa/src/main/java/org/event/herfa/service/DevisService.java
package org.event.herfa.service;


import org.event.herfa.dto.requestDTO.DevisRequestDTO;
import org.event.herfa.entity.Artisan;
import org.event.herfa.entity.Client;
import org.event.herfa.entity.Devis;
import org.event.herfa.entity.DevisStatus;
import org.event.herfa.repository.ArtisanRepository;
import org.event.herfa.repository.ClientRepository;
import org.event.herfa.repository.DevisRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.servlet.view.InternalResourceViewResolver;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class DevisService {


    private final DevisRepository devisRepository;
    private final ArtisanService artisanService;
    private final ClientService clientService;
    private final InternalResourceViewResolver defaultViewResolver;

    public DevisService(
            DevisRepository devisRepository,
            ArtisanService artisanService,
            ClientService clientService,
            InternalResourceViewResolver defaultViewResolver) {
        this.devisRepository = devisRepository;
        this.artisanService = artisanService;
        this.clientService = clientService;
        this.defaultViewResolver = defaultViewResolver;
    }


    //CRUD basiques

    public ResponseEntity<Devis> createDevis(Devis devis){
        return new ResponseEntity<>(devisRepository.save(devis),HttpStatus.CREATED);
    }

    public ResponseEntity<List<Devis>> getAllDevis(){
        List<Devis> devis = devisRepository.findAll();
        return new ResponseEntity<>(devis, HttpStatus.OK);
    }

    public ResponseEntity<Devis> getDevisById(Long id){
        Devis devis = devisRepository.findById(id).orElseThrow(() ->
        new RuntimeException("Devis not found"));
                return new ResponseEntity<>(devis, HttpStatus.OK);
    }



    public ResponseEntity<Devis> updateDevis(Long id, Devis devis){
        Devis existingDevis = devisRepository.findById(devis.getId()).orElseThrow(() ->
                new RuntimeException("Devis not found"));
        existingDevis.setDateDevis(devis.getDateDevis());
        existingDevis.setAmount(devis.getAmount());
        return new ResponseEntity<>(devisRepository.save(existingDevis),HttpStatus.OK);
    }

    public ResponseEntity<Void> deleteDevisById(Long id){
        devisRepository.deleteById(id);
        return new ResponseEntity<>(HttpStatus.NO_CONTENT);
    }

    public ResponseEntity<Void> deleteAllDevis(){
        devisRepository.deleteAll();
        return new ResponseEntity<>(HttpStatus.NO_CONTENT);
    }


    //Operations spécifiques

    public ResponseEntity<Devis> sendDevis(DevisRequestDTO devisRequestDTO) {
        System.out.println(devisRequestDTO.clientId());
        // get client by id
        Client client = clientService.getClientById(devisRequestDTO.clientId() ).getBody();

        // get artisan by is
        Artisan artisan = artisanService.findById(devisRequestDTO.artisanId());

        Devis devis = new Devis();

        devis.setAmount( devisRequestDTO.amount() );
        devis.setDateDevis(devisRequestDTO.dateDevis());
        devis.setStatus(devisRequestDTO.devisStatus());
        devis.setClient( client );
        devis.setArtisan( artisan );


        if (devis.getDateDevis() == null) {
            devis.setDateDevis(LocalDateTime.now());
        }
        return new ResponseEntity<>(devisRepository.save(devis),HttpStatus.OK);
    }

    public ResponseEntity<List<Devis>> listDevisByClient(Long clientId) {
    return new ResponseEntity<>(devisRepository.findByClient_Id(clientId), HttpStatus.OK);
    }

    public ResponseEntity<List<Devis>> listDevisByArtisan(Long artisanId) {
        return new ResponseEntity<>(devisRepository.findByArtisan_Id(artisanId), HttpStatus.OK);
    }

    public ResponseEntity<Devis> updateStatusDevis(Long id, DevisStatus status) {
        Devis existingDevis = devisRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Devis not found"));
        existingDevis.setStatus(status);
        return new ResponseEntity<>(devisRepository.save(existingDevis),HttpStatus.OK);
    }


}
